<i>
    Universidade Federal do Paraná          -           
    Programa de Pós-Graduação em Matemática <br>
    Álgebra Linear Aplicada                 -
    Trabalho Computacional                  <br>
    Professor Leonardo Silva de Lima        <br>
                                            <br>
    Aluno: Jhonatan Rodrigues Biller; jhonatan.biller@furg.br
</i>


<h2>
    <b>
        <font color='#660000'>
            Trabalho computacional: Redução de Dimensionalidade através da Decomposição por Valores Singulares e Agrupamento 
        </font>
    </b>
</h2>

&nbsp; Este trabalho realiza o agrupamento de um conjunto de dados, organizados em uma matriz $A$, em duas categorias utilizando a decomposição em valores singulares reduzida (SVD) dessa matriz. O agrupamento é efetuado pela redução da dimensionalidade da matriz decomposta. A fatoração é obtida através da iteração QR aplicada a $A^TA$. Ao final deste notebook, são apresentadas comparações entre os resultados obtidos para cada decomposição com dimensão reduzida, além de comparações com os resultados de bibliotecas Python já implementadas.

<h3>
    <b>   
        <font color='#092A44'>
            &nbsp; 1. Fundamentação Matemática
        </font>
    </b>
</h3>

<h4>
    <b>   
        <font color='#092A44'>
            &nbsp; &nbsp; 1.1 Transformações de Householder
        </font>
    </b>
</h4>

&nbsp; Uma transformação de Householder é uma transformação linear que reflete um vetor em relação a um hiperplano ortogonal a um vetor dado. Para um vetor $v$ não nulo em $\mathbb{R}^m$, a transformação de Householder é definida como:

$$ H = I-2ww^T. $$

&nbsp; Dado $x \in \mathbb{R}^m$, O vetor $w$ é dado por $w=\frac{x-y}{\|x-y\|}$ onde $y = -\text{sign}(x_1)\|x\|_{2}e_1$. Temos que $H$ é ortogonal e $Hx= \|x\|_{2} e_1 $. 


<h4>
    <b>   
        <font color='#092A44'>
            &nbsp; &nbsp; 1.2 Householder QR
        </font>
    </b>
</h4>

&nbsp;  Uma matriz retangular $A \in \mathbb{R}^{m \times n}$ pode ser fatorada como o produto de uma matriz ortogonal $Q \in \mathbb{R}^{m \times m}$ e uma matriz triangular superior $R \in \mathbb{R}^{m \times n}$:

$$ A = QR.$$

&nbsp; A fatoração QR pode ser encontrada através de transformações de Householder. Inicialmente, assuma que as matrizes de Householder $H_1$ e $H_2$ foram computadas de modo que:

$$ H_2 H_1 A = \begin{bmatrix}
 \times & \times  & \times  & \times & \times \\
0 & \times  & \times & \times  & \times  \\
0 & 0 & \textbf{x}  & \times  & \times  \\
0 & 0 & \textbf{x} &  \times & \times \\
0 & 0 & \textbf{x} &  \times & \times \\
0 & 0  &  \textbf{x} & \times  & \times  \\
\end{bmatrix}, $$

&nbsp; Observando as entradas destacadas, determina-se uma matriz de Householder $\tilde{H_3} \in R \in \mathbb{R}^{4 \times 4}$ tal que:
$$
\tilde{H_3}
\begin{bmatrix}
 \textbf{x} \\
 \textbf{x} \\
 \textbf{x} \\
 \textbf{x}
\end{bmatrix}
=\begin{bmatrix}
 \times \\
 0 \\
 0 \\
 0
\end{bmatrix},$$

&nbsp; Se $H_3=\text{diag}(I_2,\tilde{H_3})$, então:

$$ H_3H_2H_1A=
\begin{bmatrix}
 \times & \times  & \times  & \times & \times \\
0 & \times  & \times & \times  & \times  \\
0 & 0 & \times  & \times  & \times  \\
0 & 0 & 0 &  \times & \times \\
0 & 0 & 0 &  \times & \times \\
0 & 0  &  0 & \times  & \times  \\
\end{bmatrix},$$

Após $n$ passos obtemos uma matriz triangular superior $H_nH_{n-1} \cdots H_1A=R$ e escrevendo $Q=H_1 \cdots H_n$ obtemos $A=QR$. 

<h4>
    <b>   
        <font color='#092A44'>
            &nbsp; 1.3 Iteração QR
        </font>
    </b>
</h4>

&nbsp; Uma matriz quadrada $A \in \mathbb{R}^{n \times n}$ pode ser iterativamente decomposta através da iteração QR. Este método visa encontrar os autovalores de $A$. Iniciando com $A_0 = A$, para cada iteração $k=1,2,3,\dots$:

$$ A_k =  Q_kR_k,$$ 
$$ A_{k+1} =  R_kQ_k.$$ 

&nbsp; Os passos acima devem ser repetidos até que $A_k$ convirja para uma forma diagonal superior ou próxima o suficiente para os autovalores desejados. Ao final da iteração, $A_k$​ converge para uma matriz diagonal superior $D$, onde os elementos diagonais de $D$ são os autovalores de $A$.

<h4>
    <b>   
        <font color='#092A44'>
            &nbsp; 1.4 Decomposição em Valores Singulares
        </font>
    </b>
</h4>


&nbsp; Dada uma matriz $A \in \mathbb{R}^{m \times n}$ com $m \geq n$ a Decomposição em Valores Singulares (SVD) completa é expressa como:
$$ A = U \Sigma V^T,$$

onde $U \in \mathbb{R}^{m \times m}$ e $V \in \mathbb{R}^{n \times n}$ são matrizes ortogonais. As colunas de $V$ são os autovalores da matriz $A$ e a matriz $\Sigma \in \mathbb{R}^{m \times n}$ tem ao longo de sua diagonal principal os elementos $\sigma_i$​ (chamados valores singulares) que são não-negativos e ordenados em ordem decrescente.

<h4>
    <b>   
        <font color='#092A44'>
            &nbsp; 1.5 Decomposição em Valores Singulares Reduzida
        </font>
    </b>
</h4>

&nbsp; A decomposição SVD reduzida é uma forma compacta da decomposição SVD completa (1.4) e é dada por:

$$ A = U_r \Sigma_r V^T_r,$$

onde as matrizes da decomposição possuem as mesmas propriedades exceto suas dimensões: $U_r \in \mathbb{R}^{m \times r}$, $\Sigma_r \in \mathbb{R}^{r \times r}$ e $V_r \in \mathbb{R}^{n \times r}$, onde $r=\text{rank}(A)$. 

<h4>
    <b>   
        <font color='#092A44'>
            &nbsp; 1.6 Iteração QR e a matriz $A^TA$
        </font>
    </b>
</h4>

&nbsp; A matriz $A^TA$ é simétrica e semi-definida positiva. A iteração QR faz com que ela convirja para uma matriz diagonal, cujos elementos são os autovalores de $A^TA$, que por sua vez são os quadrados dos valores singulares de $A$. Já a matriz $Q_{\text{total}}$ acumulada durante a iteração QR de $A^TA$ é composta pelos autovetores de $A^TA$, que representa a matriz $V$ da decomposição SVD de $A$. 

<h4>
    <b>   
        <font color='#092A44'>
            &nbsp; 1.7 Redução de dimensionalidade SVD
        </font>
    </b>
</h4>

&nbsp; A redução de dimensionalidade na SVD é o processo de aproximar uma matriz $A \in \mathbb{R}^{m \times n}$ por uma versão de menor <i>rank</i>, mantendo apenas os $k$ maiores valores singulares e os correspondentes vetores singulares esquerdos e direitos. Dado $A = U \Sigma V^T$, a versão reduzida é dada por:

$$A_k = U_k \Sigma_k V_k^T.$$

<h4>
    <b>   
        <font color='#092A44'>
            &nbsp; 1.8 Cálculo de $A=U \Sigma V^T$ via iteração QR de $A^TA$
        </font>
    </b>
</h4>

&nbsp; Como mensionado na subseção 1.6 e 1.4, As matrizes $\Sigma$ e $V$ que compõe a SVD de $A$ são obtidas diretamente após realizar a iteração QR de $A^TA$, pois a matriz $\Sigma$ é formada pelas raízes quadradas dos valores singulares de $A^TA$ em sua diagonal princial e a matriz $V=Q_{\text{total}}$. Para obter $U$, pode-se utilizar a relação $U=AV\Sigma^{-1}$.

&nbsp; A seção a seguir apresenta a implementação computacional do problema.

<h3>
    <b>   
        <font color='#092A44'>
            &nbsp; 2. Implementação computacional
        </font>
    </b>
</h3>

<h4>
    <b>   
        <font color='#092A44'>
            &nbsp; &nbsp; 2.1 Tratamento da matriz $A$
        </font>
    </b>
</h4>

&nbsp; No bloco de código abaixo, são importadas as seguintes bibliotecas:

<ul>
    <li><b>os</b>: Para visualização do diretório atual.</li>
    <li><b>NumPy</b>: Para manipulação de arrays.</li>
    <li><b>SciPy</b>: Para normalizar as colunas da matriz original.</li>
    <li><b>pandas</b>: Para leitura e formatação do arquivo stars.txt.</li>
    <li><b>scikit-learn</b>: Para agrupar os dados.</li>
</ul>

➥ Carregando bibliotecas

In [1]:
import os
import numpy as np
import pandas as pd
from scipy.stats import zscore
from sklearn.cluster import KMeans
from sklearn.metrics import accuracy_score

➥ Obtendo o diretório atual:

In [2]:
print(os.getcwd())

/home/jhonatan/Documentos/GitHub/Algebra-Linear_Aplicada


➥ Lendo o arquivo stars.txt e armazenando na variável <b>A_completa</b>.

In [3]:
df = pd.read_csv('stars.txt', header=None)
A_completa = df.values.tolist()
A_completa = np.array(A_completa)
A_completa.shape

(5802, 19)

➥ Removendo a última coluna de <b>A_completa</b> (gabarito) e armazenando em <b>A</b>.

In [4]:
A = A_completa[:, :-1]
A.shape

(5802, 18)

➥ Aplicando o z-score nas colunas de <b>A</b>.

In [5]:
A = zscore(A, axis=0)

<h4>
    <b>   
        <font color='#092A44'>
            &nbsp; &nbsp; 2.2 Definição de funções
        </font>
    </b>
</h4>

➥ A função a seguir calcula o vetor de Householder dado $x \in \mathbb{R}^m$:

In [6]:
def vetor_householder(x):
    y =  - np.sign(x[0]) * np.linalg.norm(x) * np.eye(1,len(x))[0]
    w = x - y
    if np.linalg.norm(w) != 0:
        w /= np.linalg.norm(w)
    return w

➥ Escrevendo a função <b>decomposicao_qr</b>, responsável por obter uma fatoração $A=QR$ via transformações de Householder:

In [7]:
def decomposicao_qr(A, threshold=1e-10):
    m, n = A.shape
    Q = np.eye(m)
    R = np.copy(A).astype(float)

    for j in range(n):
        x = R[j:, j]
        w = vetor_householder(x)

        H = np.eye(m)
        H[j:, j:] -= 2 * np.outer(w, w)

        R = np.dot(H, R)

        Q = np.dot(Q, H) 

        R[np.abs(R) < threshold] = 0

    return Q, R

➥ Escrevendo a função <b>iteracao_qr</b> para obter os autovalores e autovetores de $A^TA$ e $AA^T$:

In [8]:
def iteracao_qr(A,n_it,tol=1e-10):

    Q_total = np.eye(A.shape[0])
    
    for i in range(0,n_it):
        
        Q, R    = decomposicao_qr(A)
        A       = R @ Q
        Q_total = Q_total @ Q

        soma_fora_diag = np.sum(np.abs(A - np.diag(np.diagonal(A))))
        if soma_fora_diag < tol:
            break

    autovalores = np.diag(A)
    autovetores = Q_total
    
    return autovalores, autovetores

➥ Escrevendo a função <b>reducao_ordem</b>, essa função é reponsável para obter uma aproximação de $A$ pela redução de ordem da decomposição SVD da matriz $A$:

In [9]:
def reducao_ordem(U, Sigma, VT, k):

    k      -= 1
    U_k     = U[:, :k]  
    Sigma_k = Sigma[:k, :k] 
    VT_k    = VT[:k, :] 
    A_k     = U_k @ Sigma_k @ VT_k
    
    return A_k

<h4>
    <b>   
        <font color='#092A44'>
             &nbsp; &nbsp; 2.3 Cálculo de $A=U \Sigma V^T$
        </font>
    </b>
</h4>

➥ Calculando a matriz $A^TA$:

In [10]:
ATA = A.T @ A
ATA.shape

(18, 18)

➥ Calculando os autovalores e autovetores de $A^TA$ via iteração QR e utilizando o <b>NumPy</b> para comparação dos resultados obtidos:

In [11]:
autovalores_ATA,    autovetores_ATA    = iteracao_qr(ATA,100)
autovalores_ATA_np, autovetores_ATA_np = np.linalg.eig(ATA)

print('\nAutovalores(ATA) - Obtido pela iteração QR:\n',     np.sort(autovalores_ATA)   [::-1])
print('\nAutovalores(ATA) - Obtido pela biblioteca NumPy:\n',np.sort(autovalores_ATA_np)[::-1])


Autovalores(ATA) - Obtido pela iteração QR:
 [4.77116737e+04 4.02099709e+04 5.83918387e+03 5.76621601e+03
 4.08953573e+03 7.58640003e+02 3.36090374e+01 1.11283733e+01
 6.63603707e+00 3.51928341e+00 1.96894125e+00 1.23407480e+00
 5.61467695e-01 5.10250978e-01 4.67160791e-01 4.44457139e-01
 4.03720863e-01 2.96985967e-01]

Autovalores(ATA) - Obtido pela biblioteca NumPy:
 [4.77116737e+04 4.02099709e+04 5.83937219e+03 5.76602768e+03
 4.08953573e+03 7.58640003e+02 3.36090374e+01 1.11283733e+01
 6.63603707e+00 3.51928341e+00 1.96894125e+00 1.23407480e+00
 5.61467697e-01 5.10250976e-01 4.67172071e-01 4.44445858e-01
 4.03720863e-01 2.96985967e-01]


➥ Ordenando os autovalores e autovetores de $A^TA$:

In [12]:
idx             = autovalores_ATA.argsort()[::-1]
autovalores_ATA = autovalores_ATA[idx]
autovetores_ATA = autovetores_ATA[:, idx]

➥ Construindo as matrizes $V$ e $\Sigma$ de $A$:

In [13]:
# Matriz Sigma:
valores_singulares_A = np.sqrt(autovalores_ATA)
Sigma                = np.zeros((A.shape[0], A.shape[1]))
np.fill_diagonal(Sigma, valores_singulares_A)

#Matriz V:
V  = autovetores_ATA
VT = V.T

print('Dimensões Sigma:',Sigma.shape)
print('Dimensões V:',    V.shape)

Dimensões Sigma: (5802, 18)
Dimensões V: (18, 18)


➥ Obtendo matrizes $U_{\text{np}}$, $\Sigma_{\text{np}}$ e $V_{\text{np}}^{T}$ pela biblioteca NumPy para comparações dos resultados obtidos:

In [14]:
U_np, Sigma_np, VT_np = np.linalg.svd(A, full_matrices=False)

print("diag(Sigma):\n" ,np.diag(Sigma))
print("diag(Sigma_np):\n",Sigma_np)

print("VT[0]:\n",VT[0])
print("VT_np[0]:\n",VT_np[0])


diag(Sigma):
 [218.43002014 200.52424018  76.41455273  75.93560434  63.94947795
  27.54342032   5.7973302    3.33592165   2.57605067   1.87597532
   1.40318967   1.11088919   0.74931148   0.71431854   0.68349162
   0.66667619   0.63539032   0.54496419]
diag(Sigma_np):
 [218.43002014 200.52424018  76.41578497  75.9343643   63.94947795
  27.54342032   5.7973302    3.33592165   2.57605067   1.87597532
   1.40318967   1.11088919   0.74931148   0.71431854   0.68349987
   0.66666773   0.63539032   0.54496419]
VT[0]:
 [ 0.00388597 -0.0045566   0.20337757 -0.27160096  0.20490543 -0.29079334
  0.20429353 -0.30467715  0.19820463 -0.30858524  0.18904348 -0.31135684
  0.16221051 -0.30780766  0.14796042 -0.31117205  0.13999879 -0.31169757]
VT_np[0]:
 [ 0.00388597 -0.0045566   0.20337754 -0.27160098  0.2049054  -0.29079336
  0.2042935  -0.30467717  0.1982046  -0.30858525  0.18904345 -0.31135685
  0.16221047 -0.30780768  0.14796039 -0.31117206  0.13999876 -0.31169759]


➥ Construindo a Matriz $U$ = $AV\Sigma^{-1}$:

In [15]:
U =  A @ V  @ np.linalg.pinv(Sigma)

print('U[:,0]\n',U[:,0])
print('U[:,0]\n',U_np[:,0])

print('U[:,1]\n',U[:,1])
print('U[:,1]\n',U_np[:,1])

print('U.T @ U\n',np.diag(U.T @ U)[0:17])

U[:,0]
 [-0.04529033 -0.04557697 -0.04496722 ...  0.01037127  0.01172578
 -0.00355707]
U[:,0]
 [-0.04529034 -0.04557698 -0.04496722 ...  0.01037127  0.01172578
 -0.00355707]
U[:,1]
 [ 0.04405459  0.04334411  0.04406847 ...  0.00943483  0.01145516
 -0.01688796]
U[:,1]
 [-0.04405458 -0.04334411 -0.04406846 ... -0.00943483 -0.01145517
  0.01688796]
U.T @ U
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]


➥ Calculando  $\|A-U\Sigma V^T \|_F$

In [16]:
np.linalg.norm(A - U @ Sigma @ VT)

1.4661761919142798e-12

➥ Realizando redução de ordem para $k = 2,3,4,5$:

In [17]:
A_2 = reducao_ordem(U, Sigma, VT, 2)
A_3 = reducao_ordem(U, Sigma, VT, 3)
A_4 = reducao_ordem(U, Sigma, VT, 4)
A_5 = reducao_ordem(U, Sigma, VT, 5)

print('norma(A2-A)=',np.linalg.norm(A_2-A))
print('norma(A3-A)=',np.linalg.norm(A_3-A))
print('norma(A4-A)=',np.linalg.norm(A_4-A))
print('norma(A5-A)=',np.linalg.norm(A_5-A))

norma(A2-A)= 238.16869294624158
norma(A3-A)= 128.50819195265666
norma(A4-A)= 103.32072168794676
norma(A5-A)= 70.06393882510014


➥ Armazenando a última coluna de <b>A_completa</b> (gabarito):

In [18]:
y_true = A_completa[:,-1]
y_true

array([0., 0., 0., ..., 1., 1., 1.])

➥ Realizando o agrupamento para <b>A_2</b>, <b>A_3</b>, <b>A_4</b> e <b>A_5</b>, utilizando $U$, $\Sigma$ e $V$:

In [19]:
# Agrupamento com k-means para A_2
kmeans_2 = KMeans(n_clusters=2, random_state=42)
y_pred_2 = kmeans_2.fit_predict(A_2)

# Agrupamento com k-means para A_3
kmeans_3 = KMeans(n_clusters=2, random_state=42)
y_pred_3 = kmeans_3.fit_predict(A_3)

# Agrupamento com k-means para A_4
kmeans_4 = KMeans(n_clusters=2, random_state=42)
y_pred_4 = kmeans_4.fit_predict(A_4)

# Agrupamento com k-means para A_5
kmeans_5 = KMeans(n_clusters=2, random_state=42)
y_pred_5 = kmeans_5.fit_predict(A_5)

# Comparação com o Gabarito
accuracy_2 = accuracy_score(y_true, y_pred_2)
accuracy_3 = accuracy_score(y_true, y_pred_3)
accuracy_4 = accuracy_score(y_true, y_pred_4)
accuracy_5 = accuracy_score(y_true, y_pred_5)

print("Acurácia (Dimensão 2):", 100 - accuracy_2*100)
print("Acurácia (Dimensão 3):", 100 - accuracy_3*100)
print("Acurácia (Dimensão 4):", 100 - accuracy_4*100)
print("Acurácia (Dimensão 5):", 100 - accuracy_5*100)

Acurácia (Dimensão 2): 97.88004136504654
Acurácia (Dimensão 3): 99.96552912788694
Acurácia (Dimensão 4): 99.96552912788694
Acurácia (Dimensão 5): 99.96552912788694


➥ Realizando o agrupamento para <b>A_2_np</b>, <b>A_3_np</b>, <b>A_4_np</b> e <b>A_5_np</b>, utilizando $U_{\text{np}}$, $\Sigma_{\text{np}}$ e $V_{\text{np}}$:

In [20]:
Sigma_np_novo = np.zeros((A.shape[1], A.shape[1]))
np.fill_diagonal(Sigma_np_novo, Sigma_np)

A_2_np = reducao_ordem(U_np, Sigma_np_novo, VT_np, 2)
A_3_np = reducao_ordem(U_np, Sigma_np_novo, VT_np, 3)
A_4_np = reducao_ordem(U_np, Sigma_np_novo, VT_np, 4)
A_5_np = reducao_ordem(U_np, Sigma_np_novo, VT_np, 5)

print('norma(A2-A)=',np.linalg.norm(A_2_np-A))
print('norma(A3-A)=',np.linalg.norm(A_3_np-A))
print('norma(A4-A)=',np.linalg.norm(A_4_np-A))
print('norma(A5-A)=',np.linalg.norm(A_5_np-A))

kmeans_2_np = KMeans(n_clusters=2, random_state=42)
y_pred_2_np = kmeans_2_np.fit_predict(A_2_np)

kmeans_3_np = KMeans(n_clusters=2, random_state=42)
y_pred_3_np = kmeans_3_np.fit_predict(A_3_np)

kmeans_4_np = KMeans(n_clusters=2, random_state=42)
y_pred_4_np = kmeans_4_np.fit_predict(A_4_np)

kmeans_5_np = KMeans(n_clusters=2, random_state=42)
y_pred_5_np = kmeans_5_np.fit_predict(A_5_np)

accuracy_2_np = accuracy_score(y_true, y_pred_2_np)
accuracy_3_np = accuracy_score(y_true, y_pred_3_np)
accuracy_4_np = accuracy_score(y_true, y_pred_4_np)
accuracy_5_np = accuracy_score(y_true, y_pred_5_np)

print("Acurácia (Dimensão 2):", 100 - accuracy_2_np*100)
print("Acurácia (Dimensão 3):", 100 - accuracy_3_np*100)
print("Acurácia (Dimensão 4):", 100 - accuracy_4_np*100)
print("Acurácia (Dimensão 5):", 100 - accuracy_5_np*100)

norma(A2-A)= 238.16869294624135
norma(A3-A)= 128.50819195265666
norma(A4-A)= 103.31981032726732
norma(A5-A)= 70.06393882510012
Acurácia (Dimensão 2): 97.88004136504654
Acurácia (Dimensão 3): 99.96552912788694
Acurácia (Dimensão 4): 99.96552912788694
Acurácia (Dimensão 5): 99.96552912788694


<h3>
    <b>   
        <font color='#092A44'>
            &nbsp; 3. Resultados
        </font>
    </b>
</h3>

&nbsp; Ao calcular a norma $\|A - U \Sigma V^T \|_F$ , obtém-se o valor $1,47 \times 10^{-12}$. Este valor indica que a decomposição SVD foi realizada com alta precisão, resultando em uma matriz $U \Sigma V^T$ próxima da matriz original $A$.

&nbsp; Com base nos resultados das normas $\|A_k- A \|_F$ para diferentes valores de $k$ após a decomposição SVD, pode-se concluir que à medida que o número de dimensões $k$ aumenta, a norma $\|A_k- A \|_F$ diminui, indicando que a matriz aproximada $A_k$ se torna cada vez mais próxima da matriz original $A$ conforme apresenta a Tabela 1.

<center><b>Tabela 1</b> - Relação da redução de dimensionalidade com $\| \cdot \|_F$.</center>

| Dimensão  | $\|A_k- A \|_F$  |
| --- |  --- |
| 2| 238 |
|3 | 128 | 
| 4|  103 |
| 5 | 70 |
<center>Fonte: O autor.</center>
<br>
&nbsp; A análise dos dados utilizando <i>clustering</i> com $n=2$ após a redução de dimensionalidade via SVD a Tabela 2 mostra os resultados de acurácia:

<center><b>Tabela 2</b> - Acerto do agrupamento em relação ao gabarito.</center>

| Dimensão  | Acerto (%)  |
| --- |  --- |
| 2| 97,88% |
|3 | 99,97% | 
| 4| 99,97% |
| 5 |99,97% |
<center>Fonte: O autor.</center>
<br>

&nbsp;  Os resultados indicam que a acurácia do clustering melhora significativamente ao aumentar a dimensionalidade de 2 para 3, atingindo um valor quase perfeito de 99,97%. Esse alto nível de acurácia é mantido ao continuar aumentando a dimensionalidade para 4 e 5. Portanto, concluímos que a redução de dimensionalidade via SVD é altamente eficaz, e que a partir de 3 dimensões, a acurácia do clustering se estabiliza, não havendo ganhos significativos adicionais ao aumentar ainda mais o número de dimensões. Esses resultados demonstram a robustez e a eficiência da SVD para a preparação de dados para técnicas de clustering, garantindo alta precisão mesmo em dimensões reduzidas.
